pkg/storage/mvcc.go

```diff
@@ -4810,28 +4810,32 @@ func MVCCGarbageCollect(
 
 		unsafeKey := iter.UnsafeKey()
 		implicitMeta := unsafeKey.IsValue()
-		// First check for the case of range tombstone covering keys when no
-		// metadata is available.
-		//
 		// Note that we naively can't terminate GC'ing keys loop early if we
 		// enter any of branches below, as it will update the stats under the
 		// provision that the (implicit or explicit) meta key (and thus all
 		// versions) are being removed. We had this faulty functionality at some
 		// point; it should no longer be necessary since the higher levels already
 		// make sure each individual GCRequest does bounded work.
-		if implicitMeta && meta.Deleted && !unsafeKey.Timestamp.Equal(realKeyChanged) {
-			// If we have implicit deletion meta, and realKeyChanged is not the first
-			// key in history, that means it is covered by a range tombstone (which
-			// was used to synthesize meta).
-			if unsafeKey.Timestamp.LessEq(gcKey.Timestamp) {
-				// If first object in history is at or below gcKey timestamp then we
-				// have no explicit meta and all objects are subject to deletion.
-				if ms != nil {
-					ms.Add(updateStatsOnGC(gcKey.Key, metaKeySize, metaValSize, meta,
-						realKeyChanged.WallTime))
+		//
+		// First check for the case of range tombstone covering keys when no
+		// metadata is available.
+		coveredByRangeTombstone := false
+		if implicitMeta {
+			if _, hasRange := iter.HasPointAndRange(); hasRange {
+				rangeKeys := iter.RangeKeys()
+				// For cases where all keys are covered by range tombstones, we need to
+				// remove key metadata. Otherwise, if tombstone is not on top, we should
+				// follow normal checks.
+				if _, ok := rangeKeys.FirstAtOrAbove(unsafeKey.Timestamp); ok {
```

Comment:
We should avoid unnecessarily using these operations, since they can be fairly expensive and we already have the information we need.

The issue here is that the range tombstone condition is wrong. It's meant to trigger whenever a point key is covered by a range tombstone, but `!unsafeKey.Timestamp.Equal(realKeyChanged)` does not satisfy that: `realKeyChanged` will equal `unsafeKey.Timestamp` for a point tombstone. It needs to compare `unsafeKey.Timestamp` to `meta.Timestamp`, since the synthetic range tombstone meta will always be higher than the point key in this case:

```diff
--- a/pkg/storage/mvcc.go
+++ b/pkg/storage/mvcc.go
@@ -4819,7 +4819,7 @@ func MVCCGarbageCollect(
                // versions) are being removed. We had this faulty functionality at some
                // point; it should no longer be necessary since the higher levels already
                // make sure each individual GCRequest does bounded work.
-               if implicitMeta && meta.Deleted && !unsafeKey.Timestamp.Equal(realKeyChanged) {
+               if implicitMeta && meta.Deleted && !meta.Timestamp.Equal(unsafeKey.Timestamp) {
                        // If we have implicit deletion meta, and realKeyChanged is not the first
                        // key in history, that means it is covered by a range tombstone (which
                        // was used to synthesize meta).
```

Reply:
Thanks for clarifying this.